### DISCUSSION TITLE
Is it possible to use more than one GitHub Labels for the GitHub Issue/PR by-label badges?

### DISCUSSION THREAD

**Discussion by JFQueralt** [badges/shields#5943] [Q&A] [answered]
Been looking around and couldn't find an answer to this.

This shield generates based on one label (Need: Position)
![GitHub issues](https://img.shields.io/github/issues-raw/TheIOFoundation/ProjectLockdown/Need%3A%20Position?label=Need%3A%20Position)

We would like however to be able to filter further by doing
Need: Position + Stage: Ready

Rationale: We publish issues in different stages and it would help to be able to show the actual number of issues that a "Ready".

Thoughts?

**Comment by calebcartwright** [comment]
At the moment I don't think so. If I recall correctly the entire input is utilized as a single label within the graphql query so any attempts to specify multiple labels for the route parameter would just end up as a long/odd singular label.

I think it'd be reasonable to try to support this, though figuring out a good way of handling the badge label would be tricky with an open ended set of input GH labels

**Comment by JFQueralt** [comment]
I am not acquainted with GitHub API (could have a look at it, I guess) and I'll assume that the limitation could come from there.
That said, if they allow filtering via querystring then it makes little sense that they wouldn't support it too.
If I source some time (&#129310;) I'll have a look around.

**Comment by calebcartwright** [comment]
It's a GraphQL query using the v4 API:

https://github.com/badges/shields/blob/87d8028b0ba385673ed79c3c2bee3eebda6b7c4d/services/github/github-issues.service.js#L317-L384

I'm not sure how that could best be mapped to an AND type filter on the labels off hand either

**Reply by JFQueralt** [reply]
I'd need to look more closely into the code to see if I can propose a change. Maybe this weekend as we have a code sprint in Project Lockdown (https://github.com/TheIOFoundation/ProjectLockdown)

That said, I found this [post](https://github.community/t/github-v4-graphql-label-filtering-behavior-and-or/14555) and tested the AND functionality successfully although only with 2 labels so far.
(Few quick tests didn't work out for more labels for an AND operation - Anything added to the filterBy field operates as an OR)

Well, at least it's something



{
  repository(owner: "TheIOFoundation", name: "ProjectLockdown") {
    label(name: "Stage: Not Ready") {
      issues(first: 100, after: null, filterBy: {labels: ["Need: Code"]}) {
        totalCount
        pageInfo {
          startCursor
          hasNextPage
          endCursor
        }
        edges {
          node {
            number
            title
            resourcePath
            url
            state
            labels(first: 100) {
              edges {
                node {
                  name
                }
              }
            }
          }
        }
      }
    }
  }
}

**Comment by chris48s** [comment]
I appreciate I'm scope-creeping this a bit, but here's a bigger idea..

In the github front-end, it is possible to construct fairly nuanced issue/PR queries. For example:
* `is:open is:issue label:core label:bug author:app/sentry-io` - open issues with the labels "core" and "bug" opened by sentry-io bot: https://github.com/badges/shields/issues?q=is%3Aopen+is%3Aissue+label%3Acore+label%3Abug+author%3Aapp%2Fsentry-io
* `is:pr is:closed label:npm-package review:approved author:chris48s` - closed PRs labelled "npm-package" submitted by user "chris48s"  with one or more approving reviews: https://github.com/badges/shields/pulls?q=is%3Apr+is%3Aclosed+label%3Anpm-package+review%3Aapproved+author%3Achris48s

If we're going to go down the route of extending these badges to allow more advanced queries, I think the ideal solution would actually be if we supported that same query syntax in its entirity. Then users can build a query through the GH frontend and once they're happy with it, pass the query to a badge URL in a querystring. One of the nice things about the Github V4 API (which didn't exist when these badges were first designed) is we can actually use that syntax directly in a GraphQL query. So the queries above translate to:

```gql
query {
  search(query: "repo:badges/shields is:open is:issue label:core label:bug author:app/sentry-io", type: ISSUE) {
    issueCount
  }
}
```

and

```gql
query {
  search(query: "repo:badges/shields is:pr is:closed label:npm-package review:approved author:chris48s", type: ISSUE) {
    issueCount
  }
}
```

I think rather than gradually implement that syntax one feature request at a time, how about we implement one badge:

Github Issue (with arbitrary query): `/github/issue-query?q=...`

..and then we make our existing random mashup of github issue/PR variant badges redirects to preset queries on that general badge but we say "no more presets". Then any arbitrary query you can make through the GH builder can be a badge, we put this to bed once and for all and we can collapse this mess.

![Screenshot at 2020-12-14 20-47-52](https://user-images.githubusercontent.com/6025893/102134380-52a75000-3e4e-11eb-8b9d-48cd549387aa.png)

Thoughts?

**Reply by JFQueralt** [reply]
I was originally hoping for this to be the way to go as the querystring made a lot of sense.
The problem I found and made me go to the GraphQL I posted was that I couldn't figure out how to pass labels with special characters such as spaces and ":". I tried all I could think of and the query always returned 0.

**Reply by JFQueralt** [reply]
Just tried again using all combinations I could think of and it doesn't fly.
Ideas?

query {
  search(first: 100, type: ISSUE, query: "repo:TheIOFoundation/ProjectLockdown state:closed label:'Team: Devs'") {
    issueCount
    pageInfo {
      hasNextPage
      endCursor
    }
    edges {
      node {
        ... on Issue {
          createdAt
          title
          url,
          repository {
            name
          }
        }
      }
    }
  }
}

**Reply by calebcartwright** [reply]
> Thoughts?

Yeah I think that's fair and had it in the back of my mind too as I feel like something similar came up on a different thread not too terribly long ago. It does beg a few questions for me though:

* Do the user friendly queries always map nicely 1:1 with their corresponding graphql queries?
* What should the story with the badge labels be on the new badges, maybe something generic like `count` or `query` or could we still have `issues` and `pulls`? Would our redirectors need to tack on a `?label` query param (conditionally to avoid stomping any already custom labels) to the new route?
* Does this have any potential implications on our token query hits/rate limiting?

**Reply by calebcartwright** [reply]
> Just tried again using all combinations I could think of and it doesn't fly.
Ideas?

What about this?
`query: "org:TheIOFoundation repo:ProjectLockdown state:closed label:\"Team: Devs\"`

**Comment by chris48s** [comment]
@JFQueralt The corresponding graphql query for

https://github.com/TheIOFoundation/ProjectLockdown/issues?q=is%3Aopen+is%3Aissue+label%3A%22Need%3A+Position%22+label%3A%22Stage%3A+Ready%22

would be

```gql
query {
  search(query: "repo:TheIOFoundation/ProjectLockdown is:open is:issue label:\"Need: Position\" label:\"Stage: Ready\"", type: ISSUE) {
    issueCount
  }
}
```

@calebcartwright 

> Do the user friendly queries always map nicely 1:1 with their corresponding graphql queries?

That is a good question. As far as I'm aware, yes but unfortunately the docs are not totally clear so we should test. I'd be surprised if there is any divergence though.

> What should the story with the badge labels be on the new badges

For the existing badges, we should retain the same labels. I think for the "arbitrary search query" badge I think we just have to give it a generic label like "query" or "search" and leave it to the user to pass a suitable `?label=` arg. I don't think we can attempt to generically parse something like `is:open is:issue label:core label:bug author:app/sentry-io` into a useful label.

> Would our redirectors need to tack on a `?label` query param (conditionally to avoid stomping any already custom labels) to the new route?

Yes. I'd envisage that adding a `?label=` would be part of the redirect. Another thing we could do is instead of changing the existing ones to a redirect would be to just have one shared `fetch()` function and simplify the code under the hood but keep the existing special cases serving a badge directly without issueing a redirect if we wanted (but still not add more special-cases).

> Does this have any potential implications on our token query hits/rate limiting?

As long as we only return the count, it is a single node query so one rate limit point regardless of the query.

**Reply by calebcartwright** [reply]
> As long as we only return the count, it is a single node query so one rate limit point regardless of the query.

Excellent! Maybe I just haven't spent enough with these GitHub queries but it always impresses me how much of this knowledge you seem to keep in your head :smile:

**Reply by JFQueralt** [reply]
Fantastic. "\" never crossed my mind.

Excited about the possibilities.

**Comment by chris48s** [comment]
> it always impresses me how much of this knowledge you seem to keep in your head

:books: I cheated and read the docs :laughing: 

I've made an issue for this based on the discussion: https://github.com/badges/shields/issues/5948

**Comment by chris48s** [comment, marked as answer]
We've now deployed the generic issue search badge so you can now make this badge (and many others) using a custom query:
![](https://img.shields.io/github/issues-search/TheIOFoundation/ProjectLockdown?query=is%3Aopen%20is%3Aissue%20label%3A%22Need%3A%20Position%22%20label%3A%22Stage%3A%20Ready%22) - https://img.shields.io/github/issues-search/TheIOFoundation/ProjectLockdown?query=is%3Aopen%20is%3Aissue%20label%3A%22Need%3A%20Position%22%20label%3A%22Stage%3A%20Ready%22

to customise the label:
![](https://img.shields.io/github/issues-search/TheIOFoundation/ProjectLockdown?query=is%3Aopen%20is%3Aissue%20label%3A%22Need%3A%20Position%22%20label%3A%22Stage%3A%20Ready%22&label=custom%20label) - https://img.shields.io/github/issues-search/TheIOFoundation/ProjectLockdown?query=is%3Aopen%20is%3Aissue%20label%3A%22Need%3A%20Position%22%20label%3A%22Stage%3A%20Ready%22&label=custom%20label

so I'm going to mark this as answered

**Comment by JFQueralt** [comment]
Thanks for that!

I am testing it and it works like a charm. For reference, it works even if the labels have UTF emojis on them (which was my current concern after we "beautified" them these past days.

Sample:

https://github.com/TheIOFoundation/ProjectLockdown/issues?q=is%3Aopen+is%3Aissue+label%3A%22%F0%9F%96%96+Need%3A+Position%22

[![GitHub issues](https://img.shields.io/github/issues-search/TheIOFoundation/ProjectLockdown?color=%23FF8A50&query=is%3Aopen+is%3Aissue+label%3A%22%F0%9F%96%96+Need%3A+Position%22&label=Need:%20Position)](https://github.com/TheIOFoundation/ProjectLockdown/issues?q=is%3Aopen+is%3Aissue+label%3A%22%F0%9F%96%96+Need%3A+Position%22)

Truly appreciated, this is going to be of great help.
